src/audio/igo_nr/igo_nr.c

```diff
@@ -0,0 +1,804 @@
+		idx_in += in_nch;
+	}
+
+	/* Interleave write the processed data into first channel of output buffer.
+	 * Under DEBUG mode, write the input data into second channel interleavedly.
+	 */
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]) << 8;
+
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]) << 8;
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+#endif
+
+#if CONFIG_FORMAT_S32LE
+static void igo_nr_capture_s32(struct comp_data *cd,
+			       const struct audio_stream *source,
+			       struct audio_stream *sink,
+			       int32_t src_frames,
+			       int32_t sink_frames)
+{
+	int32_t in_nch = source->channels;
+	int32_t out_nch = sink->channels;
+	int32_t i;
+	int32_t idx_in = 0;
+	int32_t idx_out = 0;
+	int32_t *x;
+	int32_t *y;
+#if CONFIG_DEBUG
+	int32_t dbg_en = cd->config.igo_params.dump_data == 1 && out_nch > 1;
+#endif
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = Q_SHIFT_RND(*x, 32, 16);
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
+
+	/* Deinterleave the source buffer and keeps the first channel data as input. */
+	for (i = 0; i < sink_frames; i++) {
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]) << 16;
+
+	/* Interleave write the processed data into first channel of output buffer.
+	 * Under DEBUG mode, write the input data into second channel interleavedly.
+	 */
+#if CONFIG_DEBUG
+		if (dbg_en) {
+			y = audio_stream_write_frag_s32(sink, idx_out + 1);
+			*y = (cd->in[cd->out_rpt]) << 16;
+		}
+#endif
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
+}
+#endif
+
+static inline int32_t set_capture_func(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sourceb;
+
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	/* The igo_nr supports S16_LE data. Format converter is needed. */
+	switch (sourceb->stream.frame_fmt) {
+#if CONFIG_FORMAT_S16LE
+	case SOF_IPC_FRAME_S16_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S16_LE");
+		cd->igo_nr_func = igo_nr_capture_s16;
+		break;
+#endif
+#if CONFIG_FORMAT_S24LE
+	case SOF_IPC_FRAME_S24_4LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S24_4LE");
+		cd->igo_nr_func = igo_nr_capture_s24;
+		break;
+#endif
+#if CONFIG_FORMAT_S32LE
+	case SOF_IPC_FRAME_S32_LE:
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S32_LE");
+		cd->igo_nr_func = igo_nr_capture_s32;
+		break;
+#endif
+	default:
+		comp_err(dev, "set_capture_func(), invalid frame_fmt");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static struct comp_dev *igo_nr_new(const struct comp_driver *drv,
+				   struct sof_ipc_comp *comp)
+{
+	struct sof_ipc_comp_process *ipc_igo_nr =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_dev *dev;
+	struct comp_data *cd = NULL;
```

Comment:
NULL assignemtn not necessary

Reply:
With this NULL assignment, the memory allocations afterwards can share the same error handler.

That is,  when `if (!cd)` holds (L243:igo_nr.c), then it goes to the last and rfree(NULL) becomes nop.

Reply:
sorry not getting this point. Is (!cd) is true, that mean cd is NULL anyway

Reply:
@fuyuntsuo none of the error paths before your alloc for `cd` jump to `err` so this NULL doesn't make a difference

Reply:
Lets fix this incrementally.

Reply:
Get it @ranj063 @cujomalainey. This issue has been resolved in #4152.